### PR TITLE
PSO Speed up

### LINKED ISSUE #218: How possible is it to run the particle optimization in parallel?
Right now- particles are evaluated iteratively.

PSO generates N amount of particles(for us, 100 alterations to the initial pose 6dof) .. at each iteration, the best ncc of each particle should be evaluated-

We loop is PSO.cpp lines 69-81

`float currentBest = host_fitness_function(gBest);

for (int i = 0; i < NUM_OF_PARTICLES*NUM_OF_DIMENSIONS; i++)
{
  float rp = getRandomClamped();
  float rg = getRandomClamped();

  velocities[i] = OMEGA * velocities[i] + c1 * rp*(pBests[i] - positions[i]) + c2 * rg*(gBest[i%NUM_OF_DIMENSIONS] - positions[i]);

  positions[i] += velocities[i];
}

OMEGA = OMEGA * 0.9f;`

According to
https://machinelearningmastery.com/a-gentle-introduction-to-particle-swarm-optimization/

Each particle's best and the whole swarm's best  are used to compute velocity ** to keep in mind for #128 

Also - our hyperparameters ( c1  c2  == 1.5 ) ... I'm reading 
[Comparing_inertia_weights_and_constriction_factors_in_particle_swarm_optimization.pdf](https://github.com/BrownBiomechanics/Autoscoper/files/13078381/Comparing_inertia_weights_and_constriction_factors_in_particle_swarm_optimization.pdf)

to see if I can figure out the rationale

### PR DESCRIPTION
**Pull request by NicerNewerCar**
* Closes #218 
* Initial results are great, cutting the optimization time to 5/6 seconds [Debug] 2/3 seconds [Release] on my machine
* Currently, I am validating the accuracy of this. I've noticed sometimes it likes to optimize to a nonideal position (ie. not in the correct position on the radiograph). Not sure if this is an issue with my changes or something pre-existing.
    * @amymmorton @jcfr If you both could play around with this PR, that may help expedite the validation process.

### CODE REVIEW

File: libautoscoper/src/PSO.hpp [hunk 1]
```diff
@@ -12,12 +12,37 @@
...
+  // Copy constructor
+  Particle(const Particle& p) {
+    ncc_val = p.ncc_val;
+    position = p.position;
+  }
+  // Default constructor
+  Particle() {
+    ncc_val = -1.0f;
+  }
+  Particle(std::vector<float> pos) {
```
Comment (jcfr) [inline libautoscoper/src/PSO.hpp:34]:
```suggestion
  Particle(const std::vector<float>& pos) {
```

**Review by jcfr — COMMENTED** [review]
Moving in that direction makes sense :ok_hand: 

Could you contribute the API changes (introduction of `Particle` and vector of `Particle`) into its own pull request.

That way the pull request introducing the threading support will be reduce to its strict minimum and the review will be easier ?

**Comment by NicerNewerCar** [comment]
See #223 

Rebase this branch once that is integrated.

**Comment by NicerNewerCar** [comment]
Rebased this with #223, Seems some of the errors I corrected over there led to this not being as fast. 223 Runs in 3-5 seconds while adding the multithreading added enough overhead that it actually slowed this down to 8-9 seconds.

**Pull request closed without merging**